pkg/pillar/zedcloud/deferred.go

```diff
@@ -68,23 +69,56 @@ type SentHandlerFunction func(
 // sentHandler is callback which will be run on successful sent
 // priorityCheckFunctions may be added to send item with matched itemType firstly
 // default function at the end of priorityCheckFunctions added to serve non-priority items
-func CreateDeferredCtx(zedcloudCtx *ZedCloudContext, sentHandler *SentHandlerFunction,
+func CreateDeferredCtx(zedcloudCtx *ZedCloudContext,
+	ps *pubsub.PubSub, agentName string, ctxName string,
+	warningTime time.Duration, errorTime time.Duration,
+	sentHandler *SentHandlerFunction,
 	priorityCheckFunctions ...TypePriorityCheckFunction) *DeferredContext {
 	// Default "accept all" priority
 	priorityCheckFunctions = append(priorityCheckFunctions,
 		func(obj interface{}) bool {
 			return true
 		})
 
-	deferredCtx := DeferredContext{
+	ctx := &DeferredContext{
 		lock:                   &sync.Mutex{},
 		Ticker:                 flextimer.NewRangeTicker(longTime1, longTime2),
 		sentHandler:            sentHandler,
 		priorityCheckFunctions: priorityCheckFunctions,
 		zedcloudCtx:            zedcloudCtx,
 	}
 
-	return &deferredCtx
+	// Start processing task
+	go ctx.processQueueTask(ps, agentName, ctxName,
+		warningTime, errorTime)
+
+	return ctx
+}
+
+func (ctx *DeferredContext) processQueueTask(ps *pubsub.PubSub,
+	agentName string, ctxName string,
+	warningTime time.Duration, errorTime time.Duration) {
+
+	log := ctx.zedcloudCtx.log
+	wdName := agentName + ctxName
+
+	stillRunning := time.NewTicker(25 * time.Second)
+	ps.StillRunning(wdName, warningTime, errorTime)
+	ps.RegisterFileWatchdog(wdName)
+
+	for {
+		select {
+		case change := <-ctx.Ticker.C:
+			start := time.Now()
+			if !ctx.HandleDeferred(change, 100*time.Millisecond, false) {
+				log.Noticef("processQueueTask: some deferred items remain to be sent")
```

Comment:
No, 1-15min

Reply:
The 1-2d is like airbag for safety as far as I understand.

Reply:
startTimer does that, BUT with these changes the current placement of that call before adding to deferredItems is incorrect.

Should move that check to after adding it, and just before the call to KickTimer, with a comment to say e.g., "Lower the retry timer while we have items to send".

Alternatively lower the timer when handleDeferred returns true. In that case one would be tempted  to move the stopTimer call to the task loop as well (to call it when the queue is empty) but need to check that this does't cause a gap is SetDeferred is running concurrently with handleDeferred.

In any case, you must fix the missing startTimer when an item is added to an empty deferredItems so that the retry happens in a timely manner.

Reply:
Sorry I misread the check as len() != 0 when it is len() == 0.
Sorry for the noise.